### PR TITLE
opnode: New Sync

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**
Semi WIP new sync + some other changes to make the system still work.

**Additional context**
Add any other context about the problem you're solving.

**Metadata**
- Fixes #[Link to Issue]

### CODE REVIEW

File: opnode/l1/source.go [hunk 1]
```diff
@@ -100,6 +101,22 @@ func (s Source) L1BlockRefByNumber(ctx context.Context, l1Num uint64) (eth.L1Blo
 	return s.l1BlockRefByNumber(ctx, new(big.Int).SetUint64(l1Num))
 }
 
+func (s Source) L1BlockRefByHash(ctx context.Context, hash common.Hash) (eth.L1BlockRef, error) {
+	header, err := s.client.HeaderByHash(ctx, hash)
+	if err != nil {
+		return eth.L1BlockRef{}, err
+	}
+	l1Num := header.Number.Uint64()
+	parentNum := l1Num
```
Comment (tynes) [inline opnode/l1/source.go:110]: Does this need the intermediate variable here? Also should it be `if parentNum == 0`?

File: opnode/rollup/driver/fake_chain.go [hunk 2]
```diff
@@ -117,6 +117,16 @@ func (m *fakeChainSource) L1BlockRefByNumber(ctx context.Context, l1Num uint64)
 	return m.l1s[m.l1reorg][l1Num], nil
 }
 
+func (m *fakeChainSource) L1BlockRefByHash(ctx context.Context, l1Hash common.Hash) (eth.L1BlockRef, error) {
+	m.log.Trace("L1BlockRefByHash", "l1Hash", l1Hash, "l1Head", m.l1head, "reorg", m.l1reorg)
```
Comment (tynes) [inline opnode/rollup/driver/fake_chain.go:121]: Does this need a mutex?

File: opnode/l1/source.go (hunk 1 above)
Comment (trianglesphere) [reply opnode/l1/source.go:110]: I'm pretty sure that this is correct. The intent is to saturate subtraction at 0 (i.e. parent block of genesis is not well formed and has 0 as the parent block number).

See https://github.com/ethereum-optimism/optimistic-specs/pull/285/files#r836779844

File: opnode/rollup/driver/fake_chain.go (hunk 2 above)
Comment (trianglesphere) [reply opnode/rollup/driver/fake_chain.go:121]: I don't believe so. There is a data race in the driver test (and only test), but I think it's related to reading state variables.

**Comment by trianglesphere** [comment]
Closing in favor of #299 (this + some changes rebased on top the new block ref format).

**Pull request closed without merging**